File: frontend/src/constants/application.ts
export type ApplicationStatus =
  | 'submitted'
  | 'org_review'
  | 'communicating'
  | 'confirmed'
  | 'offline_interview'
  | 'reserved'
  | 'waitlisted'
  | 'approved'
  | 'rejected'
  | 'withdrawn'
  | 'cancelled'
  | 'closed'

export const ApplicationStatusMap: Record<ApplicationStatus, { text: string; color: string }> = {
  submitted: { text: '已提交', color: 'blue' },
  org_review: { text: '机构审核中', color: 'gold' },
  communicating: { text: '沟通中', color: 'cyan' },
  confirmed: { text: '已确认', color: 'geekblue' },
  offline_interview: { text: '线下面签', color: 'purple' },
  reserved: { text: '预留中', color: 'orange' },
  waitlisted: { text: '候补中', color: 'volcano' },
  approved: { text: '已领养', color: 'green' },
  rejected: { text: '已拒绝', color: 'red' },
  withdrawn: { text: '已放弃', color: 'default' },
  cancelled: { text: '预留已取消', color: 'default' },
  closed: { text: '已结束', color: 'default' },
}

export type ApplicationEndReason =
  | 'final_adoption'
  | 'adopter_gave_up'
  | 'reservation_cancelled'
  | 'final_rejection'
  | 'withdrawn'
  | 'superseded'

export const ApplicationEndReasonText: Record<ApplicationEndReason, string> = {
  final_adoption: '已确认最终领养',
  adopter_gave_up: '获选人放弃，名额已释放',
  reservation_cancelled: '机构取消预留，名额已释放',
  final_rejection: '最终审核拒绝',
  withdrawn: '申请人主动撤回',
  superseded: '已有其他申请人完成领养',
}

File: backend/internal/constants/application.go
package constants

// ApplicationStatus enumerates the adoption application lifecycle.
const (
	AppStatusSubmitted        = "submitted"
	AppStatusOrgReview        = "org_review"
	AppStatusCommunicating    = "communicating"
	AppStatusConfirmed        = "confirmed"
	AppStatusOfflineInterview = "offline_interview"
	AppStatusReserved         = "reserved"
	AppStatusWaitlisted       = "waitlisted"
	AppStatusApproved         = "approved"
	AppStatusRejected         = "rejected"
	AppStatusWithdrawn        = "withdrawn"
	AppStatusCancelled        = "cancelled"
	AppStatusClosed           = "closed"
)

// ApplicationEndReason explains why a terminal application ended.
const (
	AppEndReasonFinalAdoption     = "final_adoption"
	AppEndReasonAdopterGaveUp     = "adopter_gave_up"
	AppEndReasonReservationCancel = "reservation_cancelled"
	AppEndReasonFinalRejection    = "final_rejection"
	AppEndReasonWithdrawn         = "withdrawn"
	AppEndReasonSuperseded        = "superseded"
)

// ValidApplicationStatuses returns all accepted application statuses.
func ValidApplicationStatuses() []string {
	return []string{
		AppStatusSubmitted, AppStatusOrgReview, AppStatusCommunicating,
		AppStatusConfirmed, AppStatusOfflineInterview, AppStatusReserved,
		AppStatusWaitlisted, AppStatusApproved, AppStatusRejected,
		AppStatusWithdrawn, AppStatusCancelled, AppStatusClosed,
	}
}

// IsValidApplicationStatus reports whether a status is known.
func IsValidApplicationStatus(s string) bool {
	for _, v := range ValidApplicationStatuses() {
		if v == s {
			return true
		}
	}
	return false
}

// IsActiveApplicationStatus reports whether the application can still affect adoption.
func IsActiveApplicationStatus(s string) bool {
	switch s {
	case AppStatusSubmitted, AppStatusOrgReview, AppStatusCommunicating,
		AppStatusConfirmed, AppStatusOfflineInterview, AppStatusReserved,
		AppStatusWaitlisted:
		return true
	default:
		return false
	}
}

// IsReservedApplicationStatus reports whether the applicant holds the one reservation.
func IsReservedApplicationStatus(s string) bool {
	switch s {
	case AppStatusReserved, AppStatusOrgReview, AppStatusCommunicating,
		AppStatusConfirmed, AppStatusOfflineInterview:
		return true
	default:
		return false
	}
}

// IsTerminalApplicationStatus reports whether the application has ended.
func IsTerminalApplicationStatus(s string) bool {
	return !IsActiveApplicationStatus(s)
}

// NextApplicationStatuses returns explicitly requested transitions. Automatic
// waitlist promotion is handled by the service because it updates several rows.
func NextApplicationStatuses(s string) []string {
	reviewRejections := []string{AppStatusRejected}
	switch s {
	case AppStatusSubmitted:
		return append([]string{AppStatusReserved, AppStatusOrgReview, AppStatusWithdrawn}, reviewRejections...)
	case AppStatusOrgReview:
		return append([]string{AppStatusReserved, AppStatusCommunicating, AppStatusCancelled, AppStatusWithdrawn}, reviewRejections...)
	case AppStatusCommunicating:
		return append([]string{AppStatusReserved, AppStatusConfirmed, AppStatusCancelled, AppStatusWithdrawn}, reviewRejections...)
	case AppStatusConfirmed:
		return []string{AppStatusOfflineInterview, AppStatusCancelled, AppStatusWithdrawn, AppStatusRejected}
	case AppStatusOfflineInterview:
		return []string{AppStatusApproved, AppStatusRejected, AppStatusCancelled, AppStatusWithdrawn}
	case AppStatusReserved:
		return []string{
			AppStatusOrgReview, AppStatusCommunicating, AppStatusConfirmed,
			AppStatusOfflineInterview, AppStatusApproved, AppStatusRejected,
			AppStatusCancelled, AppStatusWithdrawn,
		}
	case AppStatusWaitlisted:
		return []string{AppStatusRejected, AppStatusWithdrawn}
	default:
		return nil
	}
}
